### ISSUE TITLE
Invalid assertion in acorn-optimizer

### ISSUE THREAD

**Issue by wolfpld** [emscripten-core/emscripten#20818]
**Version of emscripten/emsdk:**
3.1.50 (047b82506d6b471873300a5e4d1e690420b582d0)

The following exception is thrown when compiling code with `-O2` or higher.

```
/home/wolf/emsdk/upstream/emscripten/tools/acorn-optimizer.js:25
    throw new Error(text);
    ^

Error
    at assert (/home/wolf/emsdk/upstream/emscripten/tools/acorn-optimizer.js:25:11)
    at /home/wolf/emsdk/upstream/emscripten/tools/acorn-optimizer.js:430:13
    at Array.forEach (<anonymous>)
    at Object.VariableDeclarator (/home/wolf/emsdk/upstream/emscripten/tools/acorn-optimizer.js:429:23)
    at c (/home/wolf/emsdk/upstream/emscripten/tools/acorn-optimizer.js:85:20)
    at recursiveWalk (/home/wolf/emsdk/upstream/emscripten/tools/acorn-optimizer.js:87:5)
    at /home/wolf/emsdk/upstream/emscripten/tools/acorn-optimizer.js:83:38
    at maybeChild (/home/wolf/emsdk/upstream/emscripten/tools/acorn-optimizer.js:47:7)
    at Array.forEach (<anonymous>)
    at visitChildren (/home/wolf/emsdk/upstream/emscripten/tools/acorn-optimizer.js:57:15)
em++: error: '/home/wolf/emsdk/node/16.20.0_64bit/bin/node /home/wolf/emsdk/upstream/emscripten/tools/acorn-optimizer.js /tmp/emscripten_temp_ctpxdg9v/x.release.jso1.js AJSDCE minifyWhitespace -o /tmp/emscripten_temp_ctpxdg9v/x.release.jso2.js' failed (returned 1)
```

https://github.com/emscripten-core/emscripten/blob/047b82506d6b471873300a5e4d1e690420b582d0/tools/acorn-optimizer.js#L428-L433

The `node.type` in the failing case is `ObjectPattern`.

**Comment by sbc100** [comment]
Thanks for the report @wolfpld.   Is this a new issue that just started happening for you?  (i.e. did it work with the same code in previous versions of emscripten?)

Do you happen for know what code is causing this in the input?    If possible perhaps you could attach the JS file that is causing the issue?   (You can use EMCC_DEBUG_SAVE=1 to have the js file saved).

**Comment by wolfpld** [comment]
The problem started to appear after integrating the Agora Video Streaming SDK (https://www.agora.io/en/) into the existing code. I also examined emscripten 3.1.48 and 3.1.47 and these versions also have issues in `acorn-optimizer.js`. There have been some fixes to the optimizer, eg #20639, so I did not look too closely at those old versions, as the code has changed anyway.

At the moment I do not have a minimal reproducible example that would show the problem. FWIW, the emscripten-generated code is optimized without problems. It is likely that the Agora SDK javascript code uses some unusual language syntax that the minimizer has trouble parsing.

**Comment by wolfpld** [comment]
The issue is not present in 3.1.35.

**Comment by wolfpld** [comment]
To reproduce the issue create the following files:

`a.cpp`:
```c++
int main()
{
	return 0;
}
```

`package.json`:
```json
{
	"devDependencies": {
		"agora-rtc-sdk-ng" : "4.19.0",
		"webpack": "^5.74.0",
		"webpack-cli": "^4.10.0"
	}
}
```

`webpack.config.js`:
```js
path = require('path');

module.exports = env => {
	return {
		entry: path.resolve(__dirname, 'index.js'),
		output: {
			filename: 'all.js',
			path: __dirname
		},
		mode: 'production',
		resolve: {
			modules: [path.resolve(__dirname, 'node_modules')],
			extensions: ['.tsx', '.ts', '.js']
		}
	}
}
```

`index.js`:
```js
import AgoraRTC from 'agora-rtc-sdk-ng'
```

Then run the following commands:
```sh
npm install
node_modules/.bin/webpack
emcc a.cpp -O2 --pre-js all.js
```

**Comment by sbc100** [comment]
Is there any chance you could attach the resulting `all.js`?

**Comment by wolfpld** [comment]
[all.js.gz](https://github.com/emscripten-core/emscripten/files/13549032/all.js.gz)

**Comment by wolfpld** [comment]
I have checked with 3.1.51-git (fb0d986c5bc0ccdbf57ce9c1ffa4bf62b0dfcc12) and I am now getting:

```
/home/wolf/emsdk/emscripten/main/tools/acorn-optimizer.js:430
          if (id.type === 'ObjectPattern') {
                 ^

TypeError: Cannot read properties of null (reading 'type')
    at traverse (/home/wolf/emsdk/emscripten/main/tools/acorn-optimizer.js:430:18)
    at traverse (/home/wolf/emsdk/emscripten/main/tools/acorn-optimizer.js:436:15)
    at Object.VariableDeclarator (/home/wolf/emsdk/emscripten/main/tools/acorn-optimizer.js:444:9)
    at c (/home/wolf/emsdk/emscripten/main/tools/acorn-optimizer.js:94:20)
    at recursiveWalk (/home/wolf/emsdk/emscripten/main/tools/acorn-optimizer.js:96:5)
    at /home/wolf/emsdk/emscripten/main/tools/acorn-optimizer.js:92:38
    at maybeChild (/home/wolf/emsdk/emscripten/main/tools/acorn-optimizer.js:56:7)
    at Array.forEach (<anonymous>)
    at visitChildren (/home/wolf/emsdk/emscripten/main/tools/acorn-optimizer.js:66:15)
    at c (/home/wolf/emsdk/emscripten/main/tools/acorn-optimizer.js:92:7)
emcc: error: '/home/wolf/emsdk/node/16.20.0_64bit/bin/node /home/wolf/emsdk/emscripten/main/tools/acorn-optimizer.js /tmp/emscripten_temp_znpa_9gt/a.out.js JSDCE minifyWhitespace -o /tmp/emscripten_temp_znpa_9gt/a.out.jso1.js' failed (returned 1)
```

Same repro.

**Comment by sbc100** [comment]
Followup fix in #20843.

I verified that your `all.js` can now run through our acorn pass.
